lua_modules/ftp/server.lua

```diff
@@ -0,0 +1,157 @@
+-- a simple ftp server
+USER = "test"
+PASS = "12345"
+local file,net,pairs,print,string,table = file,net,pairs,print,string,table
+data_fnc = nil
+ftp_data = net.createServer(net.TCP, 180)
+ftp_data:listen(20, function (s)
+  if data_fnc then data_fnc(s) end
+end)
+ftp_srv = net.createServer(net.TCP, 180)
+ftp_srv:listen(21, function(socket)
+  local s = 0
+  local cwd = "/"
+  local buf = ""
+  local t = 0
+  socket:on("receive", function(c, d)
+    a = {}
+    for i in string.gmatch(d, "([^ \r\n]+)") do
+      table.insert(a,i)
+    end
+    if(a[1] == nil or a[1] == "")then return end
+    if(s == 0 and a[1] == "USER")then
+      if(a[2] ~= USER)then
+        return c:send("530 user not found\r\n")
+      end
+      s = 1
+      return c:send("331 OK. Password required\r\n")
+    end
+    if(s == 1 and a[1] == "PASS")then
+      if(a[2] ~= PASS)then
+        return c:send("530 \r\n")
+      end
+      s = 2
+      return c:send("230 OK.\r\n")
+    end
+    if(a[1] == "CDUP")then
+      return c:send("250 OK. Current directory is "..cwd.."\r\n")
+    end
+    if(a[1] == "CWD")then
+      if(a[2] == ".")then
+        return c:send("257 \""..cwd.."\" is your current directory\r\n")
+      end
+      cwd = a[2]
+      return c:send("250 OK. Current directory is "..cwd.."\r\n")
+    end
+    if(a[1] == "PWD")then
+      return c:send("257 \""..cwd.."\" is your current directory\r\n")
+    end
+    if(a[1] == "TYPE")then
+      if(a[2] == "A")then
+        t = 0
+        return c:send("200 TYPE is now ASII\r\n")
+      end
+      if(a[2] == "I")then
+        t = 1
+        return c:send("200 TYPE is now 8-bit binary\r\n")
+      end
+      return c:send("504 Unknown TYPE")
+    end
+    if(a[1] == "MODE")then
+      if(a[2] ~= "S")then
+        return c:send("504 Only S(tream) is suported\r\n")
+      end
+      return c:send("200 S OK\r\n")
+    end
+    if(a[1] == "PASV")then
+      local _,ip = socket:getaddr()
+      local _,_,i1,i2,i3,i4 = string.find(ip,"(%d+).(%d+).(%d+).(%d+)")
+      return c:send("227 Entering Passive Mode ("..i1..","..i2..","..i3..","..i4..",0,20).\r\n")
+    end
+    if(a[1] == "LIST" or a[1] == "NLST")then
+      c:send("150 Accepted data connection\r\n")
+      data_fnc = function(sd)
+        local l = file.list();
+        for k,v in pairs(l) do
+          if(a[1] == "NLST")then
+            sd:send(k.."\r\n")
+          else
```

Comment:
As I said previously multiple sends in a single CB is a no-no. So you should do something like 
```Lua
  local resp={}
  for k,v in pairs(file.list()) do
    resp[#resp+1] = (a[1] == "NLST") and 
                      (k.."\r\n") or
                      ("+r,s"..v..",\t"..k.."\r\n")
  end
  multi_send(resp)
```
Where `multi_send(resp)` is a separate routine declared above to do the multi-send. I'll discuss this separately.   Note the `sd` used here is an upval which wll create a resource leak.